src/Microsoft.AspNetCore.Routing/CompositeEndpointDataSource.cs

```diff
@@ -18,24 +20,44 @@ namespace Microsoft.AspNetCore.Routing
     [DebuggerDisplay("{DebuggerDisplayString,nq}")]
     public sealed class CompositeEndpointDataSource : EndpointDataSource
```

Comment:
Is it possible to make this internal at this point? or will that require a bunch more things to change?

Reply:
If it was made internal then there would not be a way to register new data sources outside of UseEndpointData. And there would be no way to get all data sources, endpoints, or a change token. I think it is a useful type, both the link generator and the matcher are using it.

Reply:
well, the discussion on Friday was about how we will satisfy the following requirements:
1. A way to get all endpoints for various information things
2. A way to get all endpoints + change notifications for link generation (can be internal)
3. A way to add data sources (can be through a builder)

I still think this this is the best API for all of those things and it should just be public.